Fixes #37636 - remove usage of "@theforeman/test"

MariaAga opened the pull request:
Working towards removing theforeman/test, context: https://community.theforeman.org/t/do-we-want-to-keep-foreman-js/37644
had to do all 3 in one pr since: 
[Fixes #37636 - remove "@theforeman/test" imports] - relying on jest config having resolved foremanReact into a valid path, otherwise I would have had to use relative path instead and thats too much work

[Fixes #37637 - Test plugins from foreman core] - for some reason npm lint needs `@theforeman/test` package, so it was easier to stop using it instead of working around it. left some of the theforeman eslint packages to remove later, as this commit only deals with whats necessary to get lint to work after removing theforeman/test 

.eslintrc turned into script/lint/lint_core_config.js, no changes except for formatting

MariaAga committed:
Fixes #37636 - remove "@theforeman/test" imports

MariaAga committed:
Fixes #37637 - Test plugins from foreman core

MariaAga commented on script/npm_test_plugin.js:
```diff
@@ -0,0 +1,139 @@
+      packageHasNodeModules && packageData.scripts && packageData.scripts.lint
+    );
+  }
+  if (passedArgs[0] && passedArgs[0][0] !== '-') {
+    dirsKeys = dirsKeys.filter(dir => dir.endsWith(passedArgs[0]));
+    passedArgs.shift();
+  }
```
add comment: if the argument is `--debug` and not a plugin name `npm test:plugins katello --debug`

MariaAga commented on script/plugin_webpack_directories.js:
```diff
@@ -53,11 +66,16 @@ var getPluginDirs = stderr =>
 var packageJsonDirs = stderr =>
   pluginPath('package.json')(getPluginDirs(stderr)).map(path.dirname);
 
+// Get plugin paths with the plugin name, dont assume the plugin path will have the pluin name
```
typo pluin

MariaAga commented on webpack/jest.config.js:
```diff
@@ -0,0 +1,80 @@
+/* eslint-disable spellcheck/spell-checker */
+const fs = require('fs');
+const path = require('path');
+// const { modules: foremanJsModules } = require('@theforeman/vendor-core');
```
remove

MariaAga commented on webpack/theforeman-test.js:
```diff
@@ -0,0 +1,36 @@
+// replaces @theforeman/test.js
```
Is a duplicate?

MariaAga commented on script/lint/link-eslint-plugin.js:
```diff
@@ -0,0 +1,32 @@
+  function createSymlink() {
+    fs.symlink(sourceDir, destinationDir, 'dir', err => {
+      if (err) {
+        console.error('Error creating symlink:', err);
```
Add eslint ignore

MariaAga commented on script/lint/plugin-lint.js:
```diff
@@ -0,0 +1,30 @@
+// This script is used to lint the plugin codebase using the generic eslint configuration.
+// usage in package.json:
+//   "scripts": {
+//     "lint": "node script/lint/plugin-lint.js"
```
`"lint": "node script/lint/plugin-lint.js"` -> `"lint": "node ../foreman/script/lint/plugin-lint.js"`

MariaAga commented on script/lint/link-eslint-plugin.js:
```diff
@@ -0,0 +1,33 @@
+const fs = require('fs');
+const path = require('path');
+
+function linkEslintPlugin(runPath = process.cwd()) {
```
Check npm link

MariaAga replied:
npm link too complicated (need package.json with specific naming, deletes files sometimes) while creating a symlink is simple.

ShimShtein replied:
Please add this as a comment to the code. I feel we will be wondering who and why added this file.
Also, @ekohl, do you think it will be OK given our long-term plans for containers e.t.c.?

MariaAga commented on package.json:
```diff
@@ -29,28 +31,48 @@
     "prettier": "^1.19.1",
     "pretty-format": "26.6.2",
     "react-dnd-test-backend": "^9.4.0",
+    "react-redux-test-utils": "^0.2.0",
```
remove

MariaAga commented:
Check if github actions need to be updated 
EDIT: https://github.com/theforeman/actions/pull/66

MariaAga commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -65,12 +68,29 @@ jobs:
+        run: bundle install 
+        working-directory: ${{ github.workspace }}/projects/foreman
+      - name: Run plugin webpack dir to test
+        run: ./script/plugin_webpack_directories.rb   
+        working-directory: ${{ github.workspace }}/projects/foreman
```
Check if needed

MariaAga commented on package-exclude.json:
```diff
@@ -1,5 +1,6 @@
 {
     "EXCLUDE_NPM_PACKAGES": [
```
Add all test packages

MariaAga commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -65,12 +68,29 @@ jobs:
         run: npm install --package-lock-only --no-audit --legacy-peer-deps
         working-directory: ${{ github.workspace }}/projects/plugin
       - name: Install ${{ matrix.plugin }} npm dependencies
         run: npm ci --no-audit --legacy-peer-deps
         working-directory: ${{ github.workspace }}/projects/plugin
```
Check if need to run on foreman as well

MariaAga commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -65,12 +68,29 @@ jobs:
+      - name: Install gems
+        run: bundle install 
+        working-directory: ${{ github.workspace }}/projects/foreman
+      # - name: Run plugin webpack dir to test
+      #   run: ./script/plugin_webpack_directories.rb   
+      #   working-directory: ${{ github.workspace }}/projects/foreman
```
delete

ShimShtein replied:
I suppose it's time to delete it :)

MariaAga commented on .eslintrc:
```diff
@@ -196,6 +199,7 @@
     ],
-    "@theforeman/rules/require-ouiaid": ["error"]
+    "@theforeman/rules/require-ouiaid": ["error"],
+    "import/no-extraneous-dependencies": "off",
```
removed in the next commit

MariaAga commented on script/lint/lint_core_config.js:
```diff
@@ -0,0 +1,220 @@
+      },
+    ],
+
+    'import/no-extraneous-dependencies': 'off',
```
Check if needed

MariaAga commented on script/lint/link-eslint-plugin.js:
```diff
@@ -0,0 +1,33 @@
+const fs = require('fs');
+const path = require('path');
+
+function linkEslintPlugin(runPath = process.cwd()) {
```
Check rulesdir

MariaAga commented on script/lint/lint_generic_config.js:
```diff
@@ -0,0 +1,84 @@
+        packageDir: packageJsonDirectories,
+      },
+    ],
+    // '@foreman/custom/require-ouiaid': 'error', TODO: uncomment when the plugins are ready
```
Check

MariaAga commented on package.json:
```diff
@@ -8,13 +8,15 @@
   "scripts": {
-    "lint": "tfm-lint",
-    "lint:custom": "eslint ./webpack",
+    "lint": "eslint ./webpack ./script",
```
test if shows errors

MariaAga replied:
they do

ShimShtein commented on webpack/assets/javascripts/react_app/components/ExternalLogout/__tests__/ExternalLogout.test.js:
```diff
@@ -1,4 +1,4 @@
-import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
+import { testComponentSnapshotsWithFixtures } from 'foremanReact/common/testHelpers';
```
Why are we using `foremanReact/common/testHelpers` here and in other places we use relative path (`../../../../common/testHelpers`)?

MariaAga replied:
Honsetly, since doing find and replace `@theforeman/test'` to `foremanReact/common/testHelpers` was easier to do than manually doing a relative path for all the files

ShimShtein replied:
What do you think, should we standardize to `foremanReact/common/testHelpers` all over the place?

MariaAga replied:
I dont think we should standardize it, as both ways are right, and if we ever regret it it will be harder to move `foremanReact` back into relative. But if we find that we need it in the future it will be easy to move relative import into `foremanReact`

MariaAga commented:
to do, update `developer_docs/plugins.asciidoc`

ShimShtein commented on script/npm_test_plugin.js:
```diff
@@ -0,0 +1,140 @@
+      stdoutBuffer = lines.pop();
+    });
+
+    let stderrBuffer = `${pluginPath}: \n`;
```
Why do we prefix the stderr buffer with the plugin path, but we don't do the same for stdout?

MariaAga replied:
added in stdout

ShimShtein commented on script/npm_test_plugin.js:
```diff
@@ -0,0 +1,140 @@
+      stderrBuffer += data.toString();
+      const lines = stderrBuffer.split('\n');
+      stderrBuffer = lines.pop();
+      lines.forEach(line => console.error(line));
```
Why do we write the errors twice: the moment they appear and on close?
I don't see the same is done for stdout

MariaAga replied:
no reason, removed

MariaAga replied:
yes reason, will simplify this function instead on push

ShimShtein commented on script/npm_lint_plugins.js:
```diff
@@ -1,20 +1,83 @@
+dirsKeys.forEach(dirsKey => {
+  const pluginPath = dirs[dirsKey];
+  const pluginLintScript = pluginDefinesLint(pluginPath);
+  if (pluginLintScript?.includes('tfm-lint')) {
```
Does it mean that if the plugin does not define lint at all, this branch will be used?

MariaAga replied:
If a plugin does not have a lint script in package.json nothing will run 
```
package.json:
...
"scripts": {
    "lint":  SOMETHING
```
The branch options are either: `"lint": "tfm-lint"` or something like `"lint": "npx eslint webpack/",`

ShimShtein replied:
I think we should also use our default configuration if lint was not specified at all in the plugin.
I am thinking about a plugin that is created after we merge this feature: it will not need to do anything to get the default lint, so why they will define it in the first place?

MariaAga replied:
Changed to check if webpack folder exists as a condition for running tests and lint

ShimShtein commented on .eslintrc.js:
```diff
@@ -0,0 +1,21 @@
+  ],
+};
+
+module.exports = combinedConfig;
```
nit: empty line at the end of the file

ShimShtein reviewed (commented):
We still need to address https://github.com/theforeman/foreman/pull/10239/commits/cc50ed1455d0912c91b3aa25253c896ca9f511ad#r2045132395

MariaAga commented:
> We still need to address https://github.com/theforeman/foreman/commit/cc50ed1455d0912c91b3aa25253c896ca9f511ad#r2045132395

@ShimShtein Not sure where this is suppose to link to? as it just shows me the top of  the commit "Fixes #37639 - Use own lint instead of https://github.com/theforeman"

ofedoren commented on script/npm_lint_plugins.js:
```diff
@@ -1,20 +1,83 @@
   return packageData.scripts && packageData.scripts.lint;
 }
+if (passedArgs[0] && passedArgs[0][0] !== '-') {
+  // if the argument is --dix and not a plugin name npm lint:plugins katello --fix
```
nit:
```suggestion
  // if the argument is --fix and not a plugin name npm lint:plugins katello --fix
```

ofedoren reviewed (commented):
> @ShimShtein Not sure where this is suppose to link to? as it just shows me the top of the commit "Fixes #37639 - Use own lint instead of https://github.com/theforeman"

I think Shim meant 

> I think we should also use our default configuration if lint was not specified at all in the plugin.
I am thinking about a plugin that is created after we merge this feature: it will not need to do anything to get the default lint, so why they will define it in the first place?

ShimShtein commented on script/npm_lint_plugins.js:
```diff
@@ -1,20 +1,83 @@
 
   return packageData.scripts && packageData.scripts.lint;
 }
+if (passedArgs[0] && passedArgs[0][0] !== '-') {
```
Please add an empty line after the method declaration and before the actual script

ShimShtein reviewed (commented):
Just one small additional comment

ShimShtein commented on script/npm_lint_plugins.js:
```diff
@@ -1,20 +1,84 @@
+];
+dirsKeys.forEach(dirsKey => {
+  const pluginPath = dirs[dirsKey];
+  const pluginLintScript = pluginDefinesLint(pluginPath);
```
Shouldn't we do the same thing as we do with lint?
If a plugin defines a `test` script - run it, otherwise use this one?

MariaAga replied:
For tests, we search for a config file and merge it with ours

MariaAga commented on webpack/jest.config.js:
```diff
@@ -0,0 +1,80 @@
+  coverageDirectory: `../coverage`,
+  setupFiles: [require.resolve('jest-prop-type-error')],
+  testRegex: '.*\\.test\\.js$',
+  testPathIgnorePatterns: [
```
add .vendor

MariaAga replied:
Also to lint

MariaAga commented:
Updated webpack/jest.config.js and script/npm_test_plugin.js to ignore fixtures and gems 
change spawn to exec
and added a message for errors so when users test all plugins they will see a list at the end of which of the runs failed

MariaAga committed:
Fixes #37639 - Use own lint instead of @theforeman

MariaAga commented:
Updated to skip gems in the plugins list, and notify user about it

ShimShtein commented on script/npm_lint_plugins.js:
```diff
@@ -1,23 +1,91 @@
+var dirsKeys = [];
+var skippedDirs = [];
+var passedArgs = process.argv.slice(2);
+Object.keys(dirs).forEach(k => {
```
Can we move it to `plugin_webpack_directories`? 
something like:
``` js
const skipGems = true;
var dirs = packageJsonDirsObject(skipGems);
```
or
``` js
var dirs = developmentPackageJsonDirs();
```
We will have a single place where we identify that the plugin is a gem. If we would change this logic later, it would be much easier.

MariaAga replied:
Added filterPluginDirectories to plugin_webpack_directories.js

ShimShtein reviewed (commented):
I am getting
```
<--- Last few GCs --->

[483922:0x555dff6ad250]   484320 ms: Scavenge (reduce) 1965.9 (2083.8) -> 1965.5 (2084.1) MB, 17.4 / 0.0 ms  (average mu = 0.150, current mu = 0.030) allocation failure; 
[483922:0x555dff6ad250]   484359 ms: Scavenge (reduce) 1966.2 (2084.1) -> 1965.8 (2084.3) MB, 18.6 / 0.0 ms  (average mu = 0.150, current mu = 0.030) allocation failure; 


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x555dfbe02954 node::Abort() [/usr/bin/node]
 2: 0x555dfbc95666 node::OOMErrorHandler(char const*, bool) [/usr/bin/node]
 3: 0x555dfc009fa4 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/bin/node]
 4: 0x555dfc00a266 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/bin/node]
 5: 0x555dfc1cf299  [/usr/bin/node]
 6: 0x555dfc1d664a  [/usr/bin/node]
 7: 0x555dfc1e6702 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::internal::GarbageCollectionReason, char const*, v8::GCCallbackFlags) [/usr/bin/node]
 8: 0x555dfc1e6f89 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/bin/node]
```
when running npm test:plugins
Is it something that we expect to happen? Specifically it happens on katello tests

Also there is
```
> katello@1.0.0 lint /home/vagrant/katello
> npx eslint webpack/

Cannot read config file: /home/vagrant/katello/webpack/.eslintrc.js
Error: Could not find ./node_modules/@theforeman/vendor-dev/eslint.extends.js in /home/vagrant/foreman
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! katello@1.0.0 lint: `npx eslint webpack/`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the katello@1.0.0 lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/vagrant/.npm/_logs/2025-05-28T09_27_08_093Z-debug.log
```
when I run lint:plugins. Is it considered a normal thing?

MariaAga commented:
For katello error: 
https://github.com/Katello/katello/pull/11395
for " JavaScript heap out of memory" I added process.env.NODE_OPTIONS =
  process.env.NODE_OPTIONS || '--max-old-space-size=8096';

MariaAga commented:
rebase

ShimShtein commented:
LGTM, @theforeman/packaging , can you give it a final ack?

MariaAga commented:
added NODE_OPTIONS to the docs

evgeni commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -45,11 +48,11 @@ jobs:
+        uses: ruby/setup-ruby@v1
         with:
           ruby-version: ${{ matrix.ruby }}
+          bundler-cache: true
```
Curious, why did you add this? At this point, neither Foreman nor a Plugin is checked out yet, so bundler won't do anything, as there is no Gemfile.

MariaAga replied:
Took i from here: https://github.com/theforeman/foreman/blob/62b3353e6b557c70f4e1e51f6d2b0de48001c119/.github/workflows/foreman.yml#L82-L86

evgeni replied:
Ah yes, but there it happens after Foreman and was checked out, so there is a Gemfile
You can try doing
```yaml
      - name: "Set up Ruby ${{ matrix.ruby }}"
        uses: ruby/setup-ruby@v1
        with:
          ruby-version: ${{ matrix.ruby }}
          bundler-cache: true
          working-directory: ${{ github.workspace }}/projects/foreman
```
*after* the "Set up plugin in Foreman" step.

MariaAga replied:
If I removed it and the tests worked, does it mean it can stay removed?  or is it a cache thing to help efficiency?

evgeni replied:
If we can cache, it's preferred, as it saves a bit time/bandwidth, but not massively (NPM is slow, Ruby… usually not so)

The cool part is that if you say `bundler-cache: true`, it does two things: it calls `bundle install` *and* it caches, so you save a step (`run: bundle install`) *and* benefit from caching (a bit)

evgeni replied:
do the JS tests even boot rails? (if they don't, there is no need for the libvirt gems etc to be present)

evgeni commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -65,12 +68,26 @@ jobs:
       - name: Run ${{ matrix.plugin }} tests
-        run: npm test
-        working-directory: ${{ github.workspace }}/projects/plugin
+        run: npm run test:plugins $(echo ${{ matrix.plugin }} | awk -F'/' '{print $NF}')
```
```suggestion
        run: npm run test:plugins $PLUGIN_NAME
```

that's the same, right?

evgeni commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -65,12 +68,26 @@ jobs:
           repository: ${{ matrix.plugin }}
           path: ${{ github.workspace }}/projects/plugin
+      - name: store plugin name
+        run: echo "PLUGIN_NAME=$(echo ${{ matrix.plugin }} | awk -F'/' '{print $NF}')" >> "${GITHUB_ENV}"
```
https://github.com/theforeman/actions/blob/v0/.github/workflows/foreman_plugin_js.yml already expects the plugin name and repository to be separate inputs. would it be beneficial if we declare the matrix above as:
```
plugin:
  - name: foreman-tasks
    repo: theforeman/foreman-tasks
  - name: katello
    repo: Katello/katello
```
and use the more structured data instead of this awk line? (I am not sure using hashes like that works in GitHub actions, and would be OK if you say "Evgeni, go do your own experimental PR" ;) )

MariaAga replied:
It worked :partying_face:

evgeni commented on package.json:
```diff
@@ -60,6 +82,7 @@
     "stylelint-config-standard": "^18.0.0",
     "tabbable": "^6.2.0",
     "victory-core": "~36.8.6",
+    "victory-legend": "~36.8.6",
```
does this need to be added to the package exclude list?

MariaAga replied:
We actually never got an answer in https://github.com/theforeman/foreman/pull/10043/files#r1489694681
So I can copy from the pr and add it? but also not sure if thats the right thing?
Its what Patternfly uses for charts, and in a later pr we're going to use npm overrides to pin them https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides

evgeni replied:
So far, we've been consuming victory via `@theforeman/vendor-core` and thus the pinning here was only for development. Does this change now?

MariaAga replied:
This was added for the tests, so I'll add it to the exclude list
Once we remove `@theforeman/vendor-core`, should I open a packaging pr to remove vendor and add all its packages?

evgeni replied:
Depends :)
Once we drop vendor-core, we'll have to package all *direct* dependencies, but I'd assume all indirect ones get vendored (into the direct ones). So we'll probably package pf and pf brings victory?

MariaAga replied:
added to exclude

evgeni commented on package.json:
```diff
@@ -30,28 +32,48 @@
     "prettier": "^1.19.1",
     "pretty-format": "26.6.2",
     "react-dnd-test-backend": "^9.4.0",
+    "react-test-renderer": "^17.0.1",
```
does this need to be added to the package exclude list?

evgeni commented on package.json:
```diff
@@ -30,28 +32,48 @@
+    "@theforeman/vendor-core": "^15.0.0",
     "@types/jest": "<27.0.0",
     "argv-parse": "^1.0.1",
+    "axios-mock-adapter": "^1.1.7",
```
does this need to be added to the package exclude list?

evgeni commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -28,7 +31,7 @@ jobs:
 
   test:
-    name: ${{ matrix.plugin }} with Ruby ${{ matrix.ruby }} and Node ${{ matrix.node }}
+    name: ${{ matrix.plugin.repo }} with Ruby ${{ matrix.ruby }} and Node ${{ matrix.node }}
```
```suggestion
    name: ${{ matrix.plugin.name }} with Ruby ${{ matrix.ruby }} and Node ${{ matrix.node }}
```

Given we have the name now, I think this is nicer

evgeni commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -60,17 +64,29 @@ jobs:
         run: npm install --package-lock-only --no-audit --legacy-peer-deps
         working-directory: ${{ github.workspace }}/projects/plugin
-      - name: Install ${{ matrix.plugin }} npm dependencies
+      - name: Install ${{ matrix.plugin.repo }} npm dependencies
```
```suggestion
      - name: Install ${{ matrix.plugin.name }} npm dependencies
```
same, nicer output :)

evgeni commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -60,17 +64,29 @@ jobs:
+      - name: Install gems
+        run: bundle install 
+        working-directory: ${{ github.workspace }}/projects/foreman
+      - name: Run ${{ matrix.plugin.repo }} tests
```
```suggestion
      - name: Run ${{ matrix.plugin.name }} tests
```

evgeni reviewed (dismissed):
:package: : ACK

I left a few comments on the workflow, but nothing that can't be done later

evgeni commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -60,17 +60,35 @@ jobs:
+      - run: sudo apt-get -qq -y install build-essential libcurl4-openssl-dev zlib1g-dev libpq-dev libvirt-dev
+      - name: Install gems
+        run: bundle install 
+        working-directory: ${{ github.workspace }}/projects/foreman
```
this is taken care of by the `bundler-cache: true` above, so can be dropped

evgeni commented on .github/workflows/plugins_react_tests.yml:
```diff
@@ -60,17 +60,35 @@ jobs:
-        run: npm test
-        working-directory: ${{ github.workspace }}/projects/plugin
+      - run: sudo apt-get update
+      - run: sudo apt-get -qq -y install build-essential libcurl4-openssl-dev zlib1g-dev libpq-dev libvirt-dev
```
given `setup-ruby` with `bundler-cache: true` succeeds above, these *seem* not to be necessary.

so you can either drop them, or move before the `setup-ruby` and see what you can drop from the `BUNDLE_WITHOUT` list (realistically tho, I don't expect plugins to need libvirt integration… so… drop?)

MariaAga replied:
Dropped and it seem to work so yay
